src/hotspot/share/opto/addnode.cpp

```diff
@@ -769,6 +771,553 @@ uint AddPNode::match_edge(uint idx) const {
+    NOT_PRODUCT( COMMA _trace_tags(Compile::current()->directive()->trace_merge_loads_tags()) )
+  {}
+
+  // try merge adjacent loads, and return replace node for origin _combine
+  Node* run();
+
+private:
+  // Detect the embedding combine node is a candidate for merging loads
+  bool is_merged_load_candidate( ) const;
+  // Check other_load and load are compatible
+  bool is_compatible_load(const LoadNode* other_load, const LoadNode* load) const;
+  // From the seed load to collect load items for merging
+  void collect_merge_list(MergeLoadInfoList& merge_list, const LoadNode* load);
+  // Construct merge information item from input load
+  MergeLoadInfo* merge_load_info(LoadNode* load) const;
+  // Make the merged load and optional ReverseByte for replace
+  Node* make_merged_load(const MergeLoadInfoList& merge_list);
+
+  // Helper methods for merge loads optimization
+  static bool is_supported_load_opcode(int opcode);
+  static bool is_supported_combine_opcode(int opcode);
+  // Check the two combine operator Nodes can be reachable from one to another, and intermediate nodes has the same type
+  static bool is_reachable_combine_nodes(const Node* from, const Node* to);
+  MemoryAdjacentStatus get_adjacent_load_status(const LoadNode* first, const LoadNode* second) const;
+  // Go through ConvI2L which is unique output of input node
+  static const Node* by_pass_i2l(const Node* n);
+
+#ifndef PRODUCT
+  // Access to TraceMergeLoads tags
+  bool is_trace(TraceMergeLoads::Tag tag) const {
+    return _trace_tags.at(tag);
+  }
+
+  bool is_trace_basic() const {
+    return is_trace(TraceMergeLoads::Tag::BASIC);
+  }
+
+  bool is_trace_pointer_parsing() const {
+    return is_trace(TraceMergeLoads::Tag::POINTER_PARSING);
+  }
+
+  bool is_trace_pointer_aliasing() const {
+    return is_trace(TraceMergeLoads::Tag::POINTER_ALIASING);
+  }
+
+  bool is_trace_pointer_adjacency() const {
+    return is_trace(TraceMergeLoads::Tag::POINTER_ADJACENCY);
+  }
+
+  bool is_trace_success() const {
+    return is_trace(TraceMergeLoads::Tag::SUCCESS);
+  }
+#endif
+};
+
+bool MergePrimitiveLoads::is_supported_load_opcode(int opc) {
+  // Check for B/S/C/I
+  if (opc != Op_LoadB && opc != Op_LoadUB &&
+      opc != Op_LoadS && opc != Op_LoadUS &&
+      opc != Op_LoadI) {
+    return false;
+  }
+  return true;
+}
+
+bool MergePrimitiveLoads::is_supported_combine_opcode(int opc) {
+  return opc == Op_OrI || opc == Op_OrL;
+}
+
+// Go through ConvI2L which is unique output of input node
+const Node* MergePrimitiveLoads::by_pass_i2l(const Node* n) {
+  if (n != nullptr && n->outcnt() == 1 && n->unique_out()->Opcode() == Op_ConvI2L) {
+    return n->unique_out();
+  } else {
+    return n;
+  }
+};
+
+/*
+ * Check the _combine operator can be a candidate to trigger merge loads.
+ * The candidate load has the pattern:
+ *
+ * Load -> OrI/OrL
+ *  It has no LShift usage and the And node is optimized out in previous optimization
```

Comment:
This is a bit confusing... I thought what you wrote further up was much better, where you also had the ascii art. This here feels like a duplication of that up there. If there is anything here that is missing up there, then I would suggest that you just move it up there.